---
title: "Rare Disease Celltyping"
subtitle: "CSL Areas of Interest"
author: "Brian M. Schilder"
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: simplex
---

```{r setup, echo=FALSE}
if(!requireNamespace("MSTExplorer", quietly = TRUE)){
  remotes::install_github("neurogenomics/MSTExplorer", dependencies = TRUE)
}
if(!requireNamespace("echodata", quietly = TRUE)){
  remotes::install_github("RajLabMSSM/echodata", dependencies = TRUE)
}

knitr::opts_chunk$set(cache = TRUE,
                      echo = FALSE,
                      warning = FALSE,
                      message = FALSE) 

```

Here, we wanted to map [CSL](https://www.csl.com/) areas of interest
onto the Human Phenotype Ontology (HPO) and/or other disease ontologies
(OMIM/ORPH). This allowed us to connect our phenotype-cell type
association results to the CSL areas of interest. For a description of
our methodology, see here:

> Identification of cell type-specific gene targets underlying thousands
> of rare diseases and subtraits Kitty B. Murphy, Robert Gordon-Smith,
> Jai Chapman, Momoko Otani, Brian M. Schilder, Nathan G. Skene
> <https://www.medrxiv.org/content/10.1101/2023.02.13.23285820v1>

## Import data

### Import CSL areas of interest

First, we:

-   Gathered all diseases/phenotypes that CSL lists on their website
    and/or the Research Accelerator Initiative materials.

-   For each disease/phenotype, we assigned a Group based on whether CSL
    has expressed interest in Early Stage Partnering with external labs
    (e.g. via the RAI) or whether they are already actively pursuing
    research in these fields.

-   We then grouped each disease/phenotype into a broader Area
    ("Immunology") and a particular disease Subarea ("Dermatomyositis").

-   Next, we mapped each disease/phenotype onto a standardised HPO ID or
    OMIM/Orphanet ID.

```{r}
csl <- data.table::fread(here::here("data/CSL_areas_of_interest.tsv"))
# csl <- csl[Group=="Early Stage Partnering"]
# extract IDs from the Entry column of csl data.table with the pattern "HP:", "OMIM:", "ORPHA:" and put into a new col
csl[, ID := stringr::str_extract(Entry, "HP:\\d+|OMIM:\\d+|ORPHA:\\d+")]
# extract names from Entry column WITHOUT the ID
csl[, Name := trimws(stringr::str_remove(Entry, "HP:\\d+|OMIM:\\d+|ORPHA:\\d+"))]

MSTExplorer::create_dt(csl)
```

Next, I took all the HPO IDs and expanded this list to any HPO terms
that are descendants of the original HPO IDs. This allows us to capture
any phenotypes that are more specific than the original CSL HPO IDs.

```{r}
hpo_ids <- csl[grepl("HP:",ID)]$ID |> unique()
disease_ids <- csl[!grepl("HP:",ID)]$ID |> unique()
hpo <- KGExplorer::get_ontology("hp")
hpo_ids_list <- lapply(stats::setNames(hpo_ids,hpo_ids),
                       function(x) simona::dag_offspring(dag = hpo, 
                                                         include_self = TRUE,
                                                         term = x)) 
hpo_ids_extended <- unlist(hpo_ids_list) |> unique() 
```

### Import phenotype-cell type association results

Next, we imported the results of our phenotype-cell type association
analysis. This analysis was performed using the `MSTExplorer` package
and the results are stored in a data.table.

```{r}
results <- MSTExplorer::load_example_results()
results <- HPOExplorer::add_hpo_name(results, hpo = hpo)
results <- HPOExplorer::add_ont_lvl(results)
results <- HPOExplorer::add_ancestor(results, hpo = hpo)
results <- MSTExplorer::map_celltype(results)
MSTExplorer::add_logfc(results)
results[,effect:=estimate]

p2g <- HPOExplorer::load_phenotype_to_genes()
```

## Prioritise targets

We developed a pipeline to help filter down our results to identify the
promise promising gene targets for the most severe phenotypes. It
includes a number of different steps, including filtering on cell type
specific gene expression and the evidence supporting the causal
relationships between each gene and a phenotype.

Ultimately, it allows us to trace multi-scale disease mechanisms from
genes --\> cell types --\> phenotypes --\> diseases

```{r}
prioritise_targets_out <- MSTExplorer::prioritise_targets(
  results = results, 
  hpo = hpo,
  phenotype_to_genes = p2g,
  
  keep_deaths=NULL,
  keep_onsets=NULL,
  keep_specificity_quantiles = seq(30,40), ## NULL:70, 30-40:64 
  keep_mean_exp_quantiles = seq(30,40), ## NULL:65, 10:55
  info_content_threshold=8, ## 8:55, 5:64 
  effect_threshold=NULL, ## 1:39
  severity_score_gpt_threshold=NULL, ## 10:78, NULL:82
  symptom_intersection_threshold=.25, ## .25:57
  evidence_score_threshold=3,  ## 5:47, 4:47, 3:64
  top_n = 10, ## 5:38, 20:42, 30:45, 40:52, 50:55
  group_vars = "hpo_id")
```

Below we show a preview of the top 100 of these prioritised targets: 

```{r}
all_targets <- prioritise_targets_out$top_targets
targets <- all_targets[
  (hpo_id %in% hpo_ids_extended 
   | disease_id %in% disease_ids
   ) 
]
MSTExplorer::create_dt(head(targets,100))
```

### Validate targets

```{r ttd_check, fig.height=10, fig.width=10}
## Gene therapy only
ttd_check_out <- MSTExplorer::ttd_check(
  top_targets=prioritise_targets_out$top_targets, 
  drug_types = "Gene therapy",
  allow.cartesian = TRUE,
  show_plot = FALSE)

## All therapy types
ttd_check_all_out <- MSTExplorer::ttd_check(
  top_targets=prioritise_targets_out$top_targets,   
  allow.cartesian = TRUE,
  show_plot = FALSE)
```

To determine whether the genes prioritised by our therapeutic targets pipeline were plausible, we checked what percentage of gene therapy targets we recapitulated.
Data on therapeutic approval status was gathered from the Therapeutic Target Database (TTD; release `r Sys.Date()`).
Overall, we prioritised `r format(sum(ttd_check_out$data[failed==FALSE]$prioritised)/nrow(ttd_check_out$data[failed==FALSE])*100,digits=1)`% of all non-failed existing gene therapy targets.
A hypergeometric test confirmed that our prioritised targets were significantly enriched for non-failed gene therapy targets (`r paste0("$p=",format(ttd_check_out$ttd_hypergeo_out$nonfailed_enrichment,digits=3),"$")`).
Importantly, we did not prioritise any of the failed therapeutics (`r format(sum(ttd_check_out$data[failed==TRUE]$prioritised)/nrow(ttd_check_out$data[failed==TRUE])*100, digits=1)`%), defined as having been terminated or withdrawn from the market.
The hypergeometric test for depletion of failed targets did not reach significance (`r paste0("$p=",format(ttd_check_out$ttd_hypergeo_out$failed_depletion,digits=3),"$")`), but this is to be expected as there was only one failed gene therapy target in the TTD database.

Even when considering therapeutics of any kind, not just gene therapies, we recapitulated `r format(sum(ttd_check_all_out$data[failed==FALSE]$prioritised)/nrow(ttd_check_all_out$data[failed==FALSE])*100,digits=1)`% of the non-failed therapeutic targets and `r format(sum(ttd_check_all_out$data[failed==TRUE]$prioritised)/nrow(ttd_check_all_out$data[failed==TRUE])*100, digits=1)`% of the terminated/withdrawn therapeutic targets (n=`r nrow(ttd_check_all_out$data[failed==TRUE])`).
Here we found that our prioritised targets were significantly enriched for non-failed therapeutics (`r paste0("$p=",format(ttd_check_all_out$ttd_hypergeo_out$nonfailed_enrichment,digits=1),"$")`), and highly significantly depleted for failed therapeutics (`r paste0("$p=",format(ttd_check_all_out$ttd_hypergeo_out$failed_depletion,digits=1),"$")`).
This suggests that our multi-scale evidence-based prioritisation pipeline is capable of selectively identifying genes that are likely to be effective therapeutic targets.

```{r}
ttd_check_out$plot + patchwork::plot_annotation(title = "Gene therapy targets")
ttd_check_all_out$plot + patchwork::plot_annotation(title = "All therapy targets")
```


### Summary

Here we summarise the percentage of CSL phenotypes of interest (with HPO
IDs) included in our prioritised cell type-specific targets. We also
compute the proportion of CSL diseases (with OMIM/Orphanet IDs) that
have at least one of those phenotypes as a symptom.

```{r, message=TRUE}

message(paste0(
  length(intersect(results$hpo_id,hpo_ids_extended)),"/",
        length(unique(hpo_ids_extended)),
        " (",round(length(intersect(hpo_ids_extended,results$hpo_id))/
                  length(unique(hpo_ids_extended))*100,2),"%)",
        " CSL phenotypes",
  " covered in our phenotype-cell type association results."
))


message(paste0(
  length(intersect(all_targets$hpo_id,hpo_ids_extended)),"/",
        length(unique(hpo_ids_extended)),
        " (",round(length(intersect(hpo_ids_extended,all_targets$hpo_id))/
                  length(unique(hpo_ids_extended))*100,2),"%)",
        " CSL phenotypes (across ",length(unique(all_targets$disease_id))," diseases)",
  " covered in our prioritised targets."
))
```

### Get the top candidates per area of interest

From our list of prioritised cell type-specific gene targets, we can now
select the top targets for each CSL area of interest.

We have arbitrarily set a cutoff of 3 targets per area, but this can
easily be adjusted as needed.

```{r}
top_n <- 3
# top_targets_ancestor <- targets[!duplicated(ancestor_name),] 
csl_areas <- unique(csl[,c("ID","Area","Subarea")])
targets2 <- targets |>
  merge(csl_areas,
        all.x=TRUE,
        by.x="hpo_id",by.y="ID") |>
  merge(csl_areas,
        all.x=TRUE,
        by.x="disease_id",by.y="ID")
targets2[,Area:=data.table::fcoalesce(Area.x,Area.y)]
targets2[,Subarea:=data.table::fcoalesce(Subarea.x,Subarea.y)]
#### Select top N targets per CSL Area ####
num_cols <- sapply(targets2,is.numeric)
top_targets <- targets2[!is.na(Area), head(.SD,top_n), 
                               by=c("Area")]
# drop list columns
# save_path <- here::here("reports/top_targets_CSL.tsv")
# top_targets[,-names(top_targets)[sapply(top_targets,is.list)],with=FALSE] |>
#   data.table::fwrite(save_path, sep="\t")
# top_targets <- data.table::fread(here::here("top_targets_CSL.tsv"))

MSTExplorer::create_dt(top_targets)
```

We can also count the number of targets, genes, phenotypes and diseases
per area of interest.

```{r}
target_counts <- targets2[,list(n_targets=.N, 
                                n_genes=data.table::uniqueN(gene_symbol),
                                n_phenotypes=data.table::uniqueN(hpo_id),
                                n_diseases=data.table::uniqueN(disease_id)
              ),by=c("Area","Subarea")][!is.na(Area)]|>
  data.table::setorderv("n_targets",-1)

MSTExplorer::create_dt(target_counts)
```

## Explore top targets

Now that we have the top candidate targets for CSL disease areas, we can
explore them in more detail. For example, we can use additional
resources (ClinVar, Variant Effect Predictor, gnomad) to find out which
variants are deleterious within (or around) the gene targets. We can
also gather population-level data on the frequency of these variants,
giving us a better idea of the number of patients that may benefit from
treating this particular mechanism.

### Stroke

Let's first explore phenotypes related to "Stroke".

As an example, we're going to look specifically at the role of *COL4A1*
in stroke.

```{r}
stroke_targets <- targets2[grepl("stroke",hpo_name, ignore.case = TRUE)]
stroke_targets <- stroke_targets[gene_symbol=="COL4A1"]
```

#### ClinVar

```{r include=FALSE, eval=FALSE}
cv <- KGExplorer::get_clinvar(annotate = TRUE)
cv_gene <- cv$data[GeneSymbol=="COL4A1"]
cv_gene <- cv_gene[grepl("pathogenic",ClinicalSignificance, ignore.case = TRUE) &
                   !grepl("conflicting",ClinicalSignificance, ignore.case = TRUE),]
# txt <- paste(paste0("[",c(top_targets[1,]$disease_id,
#                           top_targets[1,]$hpo_id),"]"),collapse = "|")
cv_gene <- cv_gene[grepl(top_targets[1,]$disease_id,cv_gene$PhenotypeIDS)|
                   grepl(top_targets[1,]$hpo_id,cv_gene$PhenotypeIDS),]

length(unique(cv_gene$Name))
table(cv_gene$Type)
table(cv_gene$OriginSimple)
table(cv_gene$ClinicalSignificance)
table(cv_gene$ClinSigSimple_label)

KGExplorer::plot_clinvar(cv_gene
                         # by=c("region","Type","ClinicalSignificance"),
                         # rows = "ClinicalSignificance"
                         )
```

#### gnomad

We gathered additional data from the gnomad website on the *COL4A1* gene
(i.e.
[ENSG00000187498](https://gnomad.broadinstitute.org/gene/ENSG00000187498?dataset=gnomad_r4)).

Our goal here was to:

-   Identify confirmed pathogenic variants in *COL4A1*.

-   Check the frequency of these variants in the general population.

-   Identify a particular exon that has the highest frequency of
    pathogenic variants. The idea being that this exon could be a good
    target for therapeutic intervention via ASO-induced exon-skipping.

Additional info:

-   [Genomic coordinates of each *COL4A1* exon can be found
    here.](https://www.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000187498;r=13:110213499-110214499;t=ENST00000375820;v=rs34004222;vdb=variation;vf=813354076)

-   Exon 3 spans the following genomic coordinates:
    chr13:110213926-110214015

```{r}
#### Import ####
gnomad <- data.table::fread(here::here("data/gnomAD_v4.0.0_ENSG00000187498_2024_02_08_11_13_20.csv"),
                             check.names = TRUE)
#### Add exon information ####
db <- TxDb.Hsapiens.UCSC.hg38.knownGene::TxDb.Hsapiens.UCSC.hg38.knownGene
ex.gn <- GenomicFeatures::exonsBy(db, "gene")
ex <- ex.gn[["1282"]]
ex$exon_width <- GenomicRanges::width(ex)
# txs <- GenomicFeatures::transcriptsBy(db, by = "gene")
# tx.gn <- GenomicFeatures::exonsBy(TxDb.Hsapiens.UCSC.hg38.knownGene::TxDb.Hsapiens.UCSC.hg38.knownGene, by = "tx")
# tx <- tx.gn[["1282"]]
#### Convert to GRanges ####
gn <- echodata::dt_to_granges(gnomad,
                              chrom_col = "Chromosome", 
                              start_col = "Position",
                              style="UCSC") |>
  ## Merge gnomad with exon data
  IRanges::mergeByOverlaps(ex)
gn$ex=NULL
gn$`echodata::dt_to_granges(gnomad, chrom_col = "Chromosome", start_col = "Position", `=NULL
gn <- data.table::as.data.table(gn)
gn[,exon_variants_n:=data.table::uniqueN(gnomAD.ID),by="exon_id"]
gn[,exon_variants_perbp:=data.table::uniqueN(gnomAD.ID)/exon_width,by="exon_id"]
gn[,exon_pathovariants_perbp:=data.table::uniqueN(
  gnomAD.ID[grepl("pathogenic$",ClinVar.Clinical.Significance,ignore.case = TRUE)|cadd>=20])/exon_width,by="exon_id"]
gn[,exon_pathovariants_cumfreq:=sum(
  Allele.Frequency[grepl("pathogenic$",ClinVar.Clinical.Significance,ignore.case = TRUE)|cadd>=20]),by="exon_id"]

gn_top <-gn[exon_pathovariants_cumfreq %in% head(data.table::fsort(unique(exon_pathovariants_cumfreq),decreasing=TRUE),2)][
  grepl("pathogenic$",ClinVar.Clinical.Significance,ignore.case = TRUE)|cadd>=20,
]
gn_top[,head(.SD,1),by="exon_id",
       .SDcols = c("exon_width",
                   "exon_variants_n",
                   "exon_variants_perbp",
                   "exon_pathovariants_perbp",
                   "exon_pathovariants_cumfreq")] |>
  MSTExplorer::create_dt()
```

We found two exons with the highest cumulative frequency of pathogenic
variants in the *COL4A1* gene. Of these, exon 3 has the highest
cumulative frequency of pathogenic variants.

Follow up research into the existing literature was carried out (not
shown here), to confirm whether:

-   any similar therapies currently existing for this gene

-   exon 3 could be safely skipped

-   there are any animal models available for the homologous exons in
    this gene

```{r include=FALSE, eval=FALSE}
sort(table(gn$VEP.Annotation), decreasing = TRUE)
sort(table(gn$ClinVar.Clinical.Significance), decreasing = TRUE)

#### Find 3' UTR variants
subset(gn, 
         VEP.Annotation=="3_prime_UTR_variant" & 
         Allele.Count>10 & 
         cadd>10)

#### Find pathogenic variants with a decently highly allele frequency
common <- gn[grepl("pathogenic",ClinVar.Clinical.Significance,ignore.case = TRUE) &
                 Allele.Count>1 & 
                 cadd>20] |>
  data.table::setorderv("Allele.Frequency",-1)

#### Find exons that can be removed with no effect
gnomad[VEP.Annotation=="stop_gained" & 
         grepl("benign",ClinVar.Clinical.Significance,ignore.case = TRUE)]

#### Find splice variants that are benign
## Then find mutations that fall within the exons that are skipped within the benign splice variants
splice <- gnomad[VEP.Annotation %in% c("splice_region_variant",
                             # "intron_variant",
                             "splice_donor_variant",
                             "splice_acceptor_variant") & 
         grepl("benign",ClinVar.Clinical.Significance,ignore.case = TRUE)]
benign_transcripts <- unique(splice$Transcript)


#### Find common splicing variants that cause disease
splice <- gnomad[VEP.Annotation %in% c("splice_region_variant",
                             # "intron_variant",
                             "splice_donor_variant",
                             "splice_acceptor_variant") & 
         grepl("pathogenic",ClinVar.Clinical.Significance,ignore.case = TRUE)]
splice[Allele.Frequency>=0.001]




#### Find variants that are pathogenic in one transcript but benign in another
dat <- gnomad[grepl("benign|pathogenic",ClinVar.Clinical.Significance,ignore.case = TRUE)]
### Get the number of times a variant appears
dat <- dat[,n:=.N,by=c("Position")][n>1][,n_transcripts:=data.table::uniqueN(Transcript),by=c("Position")][]


# dat |>
  # dplyr::group_by(rsIDs) |>
  # dplyr::filter(length(unique(Transcript))>1)
# dat[data.table::uniqueN(Transcript)>1, by=c("rsIDs")]
# dat <- dat[,missense_detected:=grepl("missense",VEP.Annotation,ignore.case = TRUE),by=c("rsIDs")][missense_detected==TRUE,]

dat[,transcripts:=data.table::uniqueN(Transcript),by="Position"][transcripts>1]

unique(gnomad[,list(paste0(Transcript,":",HGVS.Consequence))][[1]])|>
  list()|> 
  utils::write.table(here::here("data/COL4A1_HGVS.txt"),sep="\t",row.names = FALSE, quote = FALSE)
```

```{r include=FALSE, eval=FALSE}
## Ensembl VEP
## https://www.ensembl.org/Homo_sapiens/Tools/VEP/Results?from=1;size=276;tl=NdzmUuJ9RE9VzgRI-9898107;to=276
vep <- data.table::fread("~/Downloads/NdzmUuJ9RE9VzgRI.txt")
```

```{r include=FALSE, eval=FALSE}
## EpiMap
## https://tiny.one/b9rxbbrv
## https://bioconductor.org/packages/release/bioc/vignettes/rhdf5/inst/doc/rhdf5_cloud_reading.html
epimap <- echotabix::query(target_path = "http://meuleman-higlass-us-west-2.altius.org/tabix/epilogos/vC.hg38.18.All_833_biosamples.S1.gz",
                 query_granges = GenomicRanges::GRanges('chr13:110150291-110150361'))

public_S3_url <- "https://storage.googleapis.com/dm-enformer/variant-scores/1000-genomes/enformer/1000G.MAF_threshold%3D0.005.13.h5"
rhdf5::h5ls(file = public_S3_url,
            s3 = TRUE,
            recursive=FALSE)
h5 <- rhdf5::H5Fopen("https://storage.googleapis.com/dm-enformer/variant-scores/1000-genomes/enformer/1000G.MAF_threshold%3D0.005.13.h5")

rhdf5::h5read(public_S3_url, 
              name = "a1", 
              index = list(1:2, 3, NULL),
              s3 = TRUE)
```

## Session info

<details>

```{r}
sessionInfo()
```

</details>
