---
title: "Rare Disease Celltyping"
subtitle: "neurodegeneration"
author: "Brian M. Schilder"
date: "<h4>Updated: <i>`r format( Sys.Date(), '%b-%d-%Y')`</i></h4>"
output: 
  html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
---

Trying to have another look through our results for anything neurodegeneration-related, either at the level of phenotype or disease.


# Import results 

```{r}
res <- MultiEWCE::load_example_results()
res <- HPOExplorer::add_disease(res)
annot <- HPOExplorer::load_phenotype_to_genes(3)
```

## Filter results

First, we do some initial filtering of the results to only get significant associations.

```{r}
res <- res[q<0.05 & symptoms.pval<0.05,]
```


## Alzheimer's

AD is not enriched for any phenotype, likely due to the small number of genes included. GWAS results are more comprehensive than this.
https://hpo.jax.org/app/browse/term/HP:0002511

```{r}
ad <- res[grepl("alzheimer",Phenotype,ignore.case = TRUE)]
nrow(ad) 
```

## Parkinson's

## Phenotypes

"Parkinson's disease" is not in the HPO, but the branch "Parkinsonism" is.
https://hpo.jax.org/app/browse/term/HP:0001300

We actually do have 1 sig phenotype here: "Parkinsonism"

```{r}
pd_phenos <- HPOExplorer::make_phenos_dataframe(ancestor="Parkinsonism")
pd1 <- res[HPO_ID %in% pd_phenos$HPO_ID ]

MultiEWCE::create_dt(pd1)
```

## Diseases

We can also search for phenotypes via "Parkinson's" as a disease.

This gives us lots of phenotypes, but many of them are not specific to PD: e.g.
*Autosomal recessive inheritance*, 
*Open mouth*, 
*Macrocephaly*,	
*Urinary urgency*

```{r}
d <- annot[grepl("parkinson",DiseaseName, ignore.case = TRUE)]
pd <- res[HPO_ID %in% d$HPO_ID] 
nrow(pd)
```

Let's try to narrow it down to very PD-specific phenotypes.

These PD-associated phenotypes only appear once in the HPO annotations. 
Thus, they must be PD-specific.

That leaves us with just 2 phenotypes.

```{r}
n_pd_terms <- length(unique(pd_phenos$HPO_ID))
pd_pheno_freqs <- sort(table(annot[HPO_ID %in% d$HPO_ID]$HPO_ID))
hist(pd_pheno_freqs, 50)
pd_specific <- pd[HPO_ID %in% names(pd_pheno_freqs)[pd_pheno_freqs<=1+n_pd_terms]]

MultiEWCE::create_dt(pd_specific)
```


We can lift this filter a bit and look for PD-associated phenotypes that only appear in 3 or less non-PD diseases. We see phenotype that can indeed be associated with PD, but are not necessarily exclusive to it: e.g.  
*Abnormal aggressive, impulsive or violent behavior*, 
*Pill-rolling tremor*, 
*"Abnormal CSF protein level* 

```{r} 
pd_specific2 <- pd[HPO_ID %in% names(pd_pheno_freqs)[pd_pheno_freqs<=3+n_pd_terms] ]

MultiEWCE::create_dt(pd_specific2)
```


## Mental deterioration

Get all the phenos from the Mental deterioration branch. This includes many forms of dementia 
 https://hpo.jax.org/app/browse/term/HP:0001268

As we've seen before, only the higher-level term "Mental deterioration" remains significant. Perhaps due to the greater number of genes in this higher level.

```{r}
phenos <- HPOExplorer::make_phenos_dataframe(ancestor="Mental deterioration")
md <- res[HPO_ID %in% phenos$HPO_ID]

MultiEWCE::create_dt(md)
```

## Atrophy/Degeneration affecting the central nervous system

HPO has the term: **Atrophy/Degeneration affecting the central nervous system** 
https://hpo.jax.org/app/browse/term/HP:0007367

In our results, we found a number of significantly enriched phenotypes
that fall under that branch:

```{r}
phenos <- HPOExplorer::make_phenos_dataframe(ancestor="Atrophy/Degeneration affecting the central nervous system")

at <- res[HPO_ID %in% phenos$HPO_ID]
MultiEWCE::create_dt(at)
```

# Multi-trait intersection


## dementia/neurodegeneration vs. diabetes/obesity

Next, let's find examples of celltype-specific enrichment results that relate to diseases/phenotypes in the categories of dementia/neurodegeneration vs. diabetes/obesity.
 
### Query results

```{r}
queries <- paste("alzheimer","parkinson","diabetes","obesity",sep = "|")
phenos <- rbind(
  HPOExplorer::make_phenos_dataframe(ancestor="Atrophy/Degeneration affecting the central nervous system"),
  HPOExplorer::make_phenos_dataframe(ancestor="Mental deterioration"),
  HPOExplorer::make_phenos_dataframe(ancestor="Diabetes mellitus"),
  HPOExplorer::make_phenos_dataframe(ancestor="Obesity")
)
res_inter <- res[grepl(queries,Phenotype,ignore.case = TRUE) | 
                 grepl(queries,DiseaseName,ignore.case = TRUE) |
                   HPO_ID %in% phenos$HPO_ID] 

message(formatC(nrow(res_inter),big.mark = ",")," results remain.")
```

### Filter results

We still have a lot of results left (500+ rows), which is a bit much to visualize.
So let's do some additional filtering.

```{r}
res_filt <- MultiEWCE::prioritise_targets(results = res_inter, 
                                          keep_tiers = NULL,
                                          severity_threshold = NULL,
                                          keep_onsets = NULL, 
                                          keep_deaths = NULL,
                                          pheno_ndiseases_threshold = NULL,
                                          symptom_p_threshold = NULL, 
                                          symptom_intersection_size_threshold = 2)
```

### Plot network

```{r, fig.height=10, fig.width=15}    
vn <- MultiEWCE::prioritise_targets_network(
  top_targets = res_filt$top_targets, 
  main = "dementia/neurodegeneration & diabetes/obesity", 
  save_path = here::here("networks/dementia_diabetes_network.html"))
```

#### [See here for full-screen version](https://neurogenomics.github.io/RareDiseasePrioritisation/networks/dementia_diabetes_network.html)

# Session info 

<details>
```{r}
sessionInfo()
```
</details>
